### ISSUE TITLE
Expand Functionality

### ISSUE THREAD

**Issue by iwitilfcuzlk** [Tyrrrz/DiscordChatExporter#271]
Now the software is very cool made, very often it helps and is used by me. The only thing in private messages opposite the correspondence is to show the date of the last message. And the latest correspondence so that you are always at the top.

I would also like to add the function of collecting links to user avatars (https://cdn.discordapp.com/avatars/158283238161383424/4f929df340856c411d68955f3df2ba92.png?size=128), user nicknames, their custom status and game status on the server that is selected from the left in software. It would be very useful for statistics. I’ve been looking for such software for a long time, but I haven’t found it, everywhere they only collect messages from chats.

**Comment by Tyrrrz** [comment]
Hi there.
The primary goal of this software is to export and archive Discord chat logs while retaining the original look and feel that the users are familiar with. There are other formats, like JSON and TXT that provide different options on the spectrum between file size and presentation.

>The only thing in private messages opposite the correspondence is to show the date of the last message. And the latest correspondence so that you are always at the top.

I'm not exactly sure what you mean by that. Do you mean that you want to see date of last message when you browse channels?

**Comment by iwitilfcuzlk** [comment]
> Hi there.
> The primary goal of this software is to export and archive Discord chat logs while retaining the original look and feel that the users are familiar with. There are other formats, like JSON and TXT that provide different options on the spectrum between file size and presentation.

Yes, but doing the parser as separate software would be weird. And since this is just an additional export window next to the existing one, you can choose to collect messages and options for it, or use a different tab in the settings for collecting. users and options to it. I saw that not only did I need such functionality because I was looking at a lot of software on github, but it simply does not exist.

> I'm not exactly sure what you mean by that. Do you mean that you want to see date of last message when you browse channels?

Make sorting in private messages as in the discord itself. And next to the personal chat is the date of the last message. Discord puts correspondence at the very top, if it is most relevant, and in the software the messages go in their own order.

**Comment by Tyrrrz** [comment]
This sounds feasible from the backend side. Need someone to come up with UI mockup that includes both the last timestamp in the UI as well as sorting. Take into account the checkmark that appears when a channel is selected as it may impose invisible margins.

**Comment by andrewkolos** [comment]
[The channel object gives a last message id.](https://discord.com/developers/docs/resources/channel#channel-object). We could then query the Discord API for that message and look at the `timestamp`. This would slow down initial load of the GUI by a about roundtrip. Would this be acceptable? Is there perhaps a way to avoid the roundtrip I don't know about?

**Comment by Tyrrrz** [comment]
The ID is actually a timestamp in itself. We currently have a method to encode a timestamp into ID, but we can make an opposite one as well. 

Here's an example: https://github.com/discord-net/Discord.Net/blob/ff0fea98a65d907fbce07856f1a9ef4aebb9108b/src/Discord.Net.Core/Utils/SnowflakeUtils.cs#L17-L18

**Comment by andrewkolos** [comment]
That's neat I looked into Snowflakes a bit more--pretty interesting stuff. I very briefly tried to implement the sorting, but one challenge is that the front-end declaratively sorts entries (see [here in RootView.xaml](https://github.com/Tyrrrz/DiscordChatExporter/blob/949c9d3f1e767796ea2f7b987d03feea1c14c186/DiscordChatExporter.Gui/Views/RootView.xaml#L32)). I discovered this when I tried a very simple approach--sort DM channels in the codebehind [(in RootViewModel.cs](https://github.com/andrewkolos/DiscordChatExporter/blob/62b064c65dc5a3b0bdba435d44640023a60295c1/DiscordChatExporter.Gui/ViewModels/RootViewModel.cs#L165)). The `SortDescription`s override any sorting of the data as it is retrieved (makes sense).

If I get rid of the declarative sorting stuff (`<CollectionViewSource.SortDescriptions>` tag), then the DM sorting works, but then the groups are no longer sorted alphabetically, which makes sense (e.g. Groups come after Private). If I just remove the `<componentModel:SortDescription Direction="Ascending" PropertyName="Name" />` and keep the `Category` line above, then the category sorting comes back, but then the DMs are not sorted at all. I am not sure what's going on here and don't want to spend much time digging at the moment, but I figure someone else might happen to know a good way to address this.

**Comment by Tyrrrz** [comment]
The sorting should happen in the view layer, ideally.
The `SortDescription` objects are composable. The way it works now is that it first sorts by category, then by channel name. I guess the idea is to add another `SortDescription` that sorts by last message date. I'm not sure if I want that though, because this might make finding channels harder. Just showing the last message date (ideally in a humanized format, like "3 days ago") should be enough.

**Comment by Tyrrrz** [comment]
Related: #428
